File: tools/binding_compliance/conformance/families/update_rejection.py
"""Negative runtime proof for default update APIs, distinct from successful service fetches."""

import json
from functools import partial

from ..coverage import CoveragePredicate, FamilyCoveragePolicy

SPECS = {
    "latest": (
        ("GithubClient", "get_latest_release"),
        (
            "get_latest_release",
            "getLatestRelease",
            "checkForUpdates",
            "github_check_for_updates",
        ),
    ),
    "all": (
        ("GithubClient", "get_all_releases"),
        ("get_all_releases", "getAllReleases"),
    ),
    "notification": (
        ("check_app_notification",),
        ("check_app_notification", "checkAppNotification"),
    ),
    "metadata": (("GithubClient",), ("repo_url",)),
}


def _negative(operation, value):
    """Only exact caller/request-builder rejection proves this negative execution path."""
    if operation == "metadata":
        return value == {"repoUrl": "https://github.com/conformance/fixture"}
    return (
        value
        == {
            "boundary": "caller-validation"
            if operation == "notification"
            else "request-builder",
            "error": "invalid-installed-version"
            if operation == "notification"
            else "builder-error",
            "requestBuilt": False,
        }
        and value["requestBuilt"] is False
    )


UPDATE_REJECTION_COVERAGE_POLICY = FamilyCoveragePolicy(
    "update-rejection",
    tuple(
        CoveragePredicate(
            id=f"update-rejection.{operation}",
            capability_id=f"update-rejection.{operation}",
            action=f"update-rejection.{operation}",
            observation_family="values"
            if operation == "metadata"
            else "negative-runtime",
            rust_symbols=symbols,
            runtime_operations=operations,
            matches=partial(_negative, operation),
        )
        for operation, (symbols, operations) in SPECS.items()
    ),
)


def validate_update_rejection_pack(document, root):
    """Admit only a synthetic invalid header or invalid installed-version fixture."""
    paths = []
    fixture_root = (root / document["fixtureRoot"]).resolve()
    for scenario in document["scenarios"]:
        operation = scenario["action"].removeprefix("update-rejection.")
        reference = scenario["input"].get("fixtureRef")
        if (
            operation not in SPECS
            or scenario["input"] != {"fixtureRef": reference}
            or scenario["fixtureRefs"] != [reference]
        ):
            raise ValueError("invalid update rejection input")
        path = (fixture_root / document["fixtures"][reference]).resolve()
        if not path.is_relative_to(fixture_root):
            raise ValueError("update rejection fixture escapes root")
        expected_input = {
            "operation": operation,
            "owner": "conformance",
            "repo": "fixture",
            "token": "synthetic\ninvalid",
            "currentVersion": "1.0.0",
            "invalidVersion": "invalid",
        }
        if json.loads(
            path.read_text(encoding="utf-8")
        ) != expected_input or not _negative(operation, scenario["expected"]):
            raise ValueError(
                "update rejection must use synthetic pre-transport failures"
            )
        paths.append(path)
    return tuple(paths)

File: business-logic/classic-scanlog-core/tests/semantic_conformance/update_rejection.rs
//! Negative runtime paths reject malformed input before an HTTP request can be sent.
#![allow(deprecated)] // The legacy latest-release entry point remains an explicit tracked obligation.

use super::{RunnerResult, invalid, text};
use classic_update_core::{GithubClient, UpdateError};
use serde_json::{Value, json};

/// Execute actual core entry points and distinguish builder rejection from transport failure.
pub(super) fn observe(fixture: &Value) -> RunnerResult<Value> {
    if fixture["token"] != "synthetic\ninvalid" {
        return Err(invalid("only synthetic malformed credentials are permitted").into());
    }
    let owner = text(&fixture["owner"])?;
    let repo = text(&fixture["repo"])?;
    let operation = text(&fixture["operation"])?;
    if operation == "metadata" {
        let client = GithubClient::with_token(&owner, &repo, Some(text(&fixture["token"])?))?;
        return Ok(json!({"repoUrl": client.repo_url()}));
    }
    let result = classic_shared_core::get_runtime().block_on(async {
        if operation == "notification" {
            classic_update_core::notification::check_app_notification(
                &owner,
                &repo,
                &text(&fixture["invalidVersion"]).expect("validated string"),
            )
            .await
            .map(|_| ())
        } else {
            let client = GithubClient::with_token(
                &owner,
                &repo,
                Some(text(&fixture["token"]).expect("validated string")),
            )?;
            if operation == "latest" {
                client.get_latest_release().await.map(|_| ())
            } else {
                client.get_all_releases(false, false).await.map(|_| ())
            }
        }
    });
    match result {
        Err(UpdateError::HttpError(error)) if operation != "notification" && error.is_builder() => {
            Ok(
                json!({"boundary": "request-builder", "error": "builder-error", "requestBuilt": false}),
            )
        }
        Err(UpdateError::NotificationInstalledVersionParse { .. })
            if operation == "notification" =>
        {
            Ok(
                json!({"boundary": "caller-validation", "error": "invalid-installed-version", "requestBuilt": false}),
            )
        }
        _ => Err(
            invalid("expected pre-transport rejection, not success or a network failure").into(),
        ),
    }
}
